File: java/JavaStreams/src/main/java/pl/sebcel/livecoding/javastreams/employeeservice/EmployeeService.java
package pl.sebcel.livecoding.javastreams.employeeservice;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class EmployeeService {

	public Map<String, List<Employee>> getMostFrequentEmployeesPerCompany(List<Company> companies) {
		validateInput(companies);
		return companies.stream()
				.collect(Collectors.toMap(
						Company::name,
						c -> getMostFrequentEmployees(c.departments()
								.stream()
								.flatMap(d -> d.employees().stream())
								))
						);
	}
	
	private List<Employee> getMostFrequentEmployees(Stream<Employee> employees) {
		Map<Employee, Integer> counts = employees.collect(Collectors.toMap(Function.identity(), _ -> 1, Integer::sum));
		int max = counts.entrySet().stream().map(Map.Entry::getValue).max(Integer::compare).orElseThrow();
		return counts.entrySet().stream().filter(e -> e.getValue() == max).map(Map.Entry::getKey).toList();
	}
	
	private void validateInput(List<Company> companies) {
		Objects.requireNonNull(companies);
		for (Company c : companies) {
			Objects.requireNonNull(c.departments());
			for (Department d : c.departments()) {
				Objects.requireNonNull(d.employees());
			}
		}
	}
}

File: csharp/LINQtoObjects/EmployeeService/EmployeeService.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace LiveCodingExercises.LINQtoObjects.EmployeeService
{
    public class EmployeeService
    {
        public IDictionary<string, IEnumerable<Employee>> GetMostFrequentEmployeesPerCompany(IEnumerable<Company> companies)
        {
            ValidateInput(companies);
            return companies.ToDictionary(c => c.Name, c => GetMostFrequentEmployeesPerSingleCompany(c.Departments.SelectMany(d => d.Employees)));
        }

        private void ValidateInput(IEnumerable<Company> companies)
        {
            ArgumentNullException.ThrowIfNull(companies);
            foreach (Company c in companies)
            {
                ArgumentNullException.ThrowIfNull(c.Departments);
                foreach (Department d in c.Departments)
                {
                    ArgumentNullException.ThrowIfNull(d.Employees);
                }
            }
        }
        
        private IEnumerable<Employee> GetMostFrequentEmployeesPerSingleCompany(IEnumerable<Employee> employees)
        {
            IDictionary<Employee, int> counts = employees.GroupBy(e => e).ToDictionary(e => e.Key, e => e.Count());
            int max = counts.Values.Max();
            return counts.Where(e => e.Value == max).Select(e => e.Key).ToList();
        }
    }
}
